File: openstudio-simulation/lib/openstudio_simulation/runner.rb
require 'fileutils'

module OpenStudioSimulation
  # Simulation execution — the SDK+CLI recipe (pure `openstudio` gem + a
  # pluggable execution backend; no measures, no openstudio-standards, no NECB).
  #
  # This module owns the BACKEND-AGNOSTIC parts: attach weather, prepare the run
  # directory (sizing flags, run period, in.osm + in.osw), and parse results from
  # the attached SQL. The actual EnergyPlus invocation is delegated to a Backend
  # (see backends.rb) — Local (the `openstudio` CLI) by default, or a remote seam.
  module Runner
    module_function

    # @return [Boolean] is the `openstudio` CLI runnable? (convenience probe;
    #   the Local backend runs its own check before executing)
    #
    # Delegates rather than duplicating: this probe and Local's drifted apart
    # once already, and both carried the same `> /dev/null` POSIX-ism that made
    # them answer "no CLI" on Windows.
    def openstudio_cli?
      @openstudio_cli = Local.new.openstudio_cli? if @openstudio_cli.nil?
      @openstudio_cli
    end

    # The process-wide default backend.
    #
    # The umbrella calls run_energyplus! at ~8 sites without a `backend:`, so a
    # process-wide default is what lets `--backend remote` reach all of them
    # without threading a parameter through performance_compliance and every
    # phase between. A per-call `backend:` still wins.
    #
    # Not thread-safe by design and not meant to be: it is set once at startup
    # from a CLI flag. Forked children (the sweep script) must set their own.
    def default_backend
      @default_backend ||= Local.new
    end

    def default_backend=(backend)
      @default_backend = backend
    end

    # Attach an EPW + its design days (required before any sizing run).
    #
    # Design days are REPLACED, not appended (a model that already carries
    # design days — every legacy archetype does — would otherwise size on
    # duplicates), and filtered to the 99.6% heating / 0.4% cooling extremes,
    # the same convention as legacy model_add_design_days_and_weather_file.
    # A full DDY carries ~40 entries including monthly and shoulder-season
    # cooling days; sizing plant equipment on those breaks E+'s cooling-tower
    # UA autosizing ("Bad starting values for UA" — found by the LargeOffice
    # archetype, the only proposed with a tower).
    def attach_weather!(model, epw:, ddy:)
      # Absolute path: the model's WeatherFile object stores this string, and
      # the CLI later resolves it from the RUN directory — a caller-relative
      # epw ('fixtures/x.epw') would not be found there. Found by running the
      # umbrella README quick-start verbatim.
      epw_file = OpenStudio::EpwFile.new(OpenStudio::Path.new(File.expand_path(epw)))
      OpenStudio::Model::WeatherFile.setWeatherFile(model, epw_file)
      workspace = OpenStudio::EnergyPlus.loadAndTranslateIdf(ddy)
      raise(ArgumentError, "could not parse design days from #{ddy}") if workspace.empty?

      model.getDesignDays.each(&:remove)
      # legacy model_set_design_days default list — NOT a bare /.4%/, which
      # would also pull the MONTHLY .4% days (a January cooling day's ~2C
      # wet-bulb is what actually breaks the tower UA solve)
      keep = [/Htg 99.6. Condns DB/, /Clg .4% Condns DB=>MWB/, /Clg 0.4% Condns DB=>MCWB/, /Clg .4. Condns WB=>MDB/]
      extremes = workspace.get.getDesignDays.select { |dd| keep.any? { |re| dd.nameString =~ re } }
      extremes = workspace.get.getDesignDays if extremes.empty? # odd DDY: keep everything rather than none
      extremes.each { |dd| model.addObject(dd.clone) }
      raise(ArgumentError, "no design days found in #{ddy}") if model.getDesignDays.empty?

      model
    end

    # Prepare the run directory, run EnergyPlus via the chosen backend, and
    # attach the result SQL to the model.
    #
    # Flow: model-prep (sizing flags, run period, save in.osm + in.osw) ->
    # backend.execute(dir) -> attach dir/run/eplusout.sql to the model.
    #
    # @param sizing_only [Boolean] design-day sizing run only
    # @param run_period [Hash, nil] { begin_month:, begin_day:, end_month:, end_day: }
    #   override for the weather run (tests use one week; code compliance is annual)
    # @param backend [Backend, nil] execution backend — nil takes
    #   `default_backend` (Local unless it has been set); swap
    #   for Remote (or any Backend) to run elsewhere without changing this method
    # @return [String] the run directory (contains eplusout.err / eplusout.sql)
    def run_energyplus!(model, dir, sizing_only: false, run_period: nil, backend: nil)
      FileUtils.mkdir_p(dir)
      sim = model.getSimulationControl
      sim.setDoZoneSizingCalculation(true)
      sim.setDoSystemSizingCalculation(true)
      sim.setDoPlantSizingCalculation(true)
      sim.setRunSimulationforSizingPeriods(true)
      sim.setRunSimulationforWeatherFileRunPeriods(!sizing_only)
      if !sizing_only && run_period
        rp = model.getRunPeriod
        rp.setBeginMonth(run_period[:begin_month])
        rp.setBeginDayOfMonth(run_period[:begin_day])
        rp.setEndMonth(run_period[:end_month])
        rp.setEndDayOfMonth(run_period[:end_day])
      end
      model.save("#{dir}/in.osm", true)
      osw = OpenStudio::WorkflowJSON.new
      # Absolute seed path: the CLI resolves a relative seed against the OSW's
      # own directory, so a caller-relative `dir` ('runs/x') would double up
      # ('runs/x/runs/x/in.osm') and fail. Found by running the README
      # quick-start verbatim.
      osw.setSeedFile(File.expand_path("#{dir}/in.osm"))
      osw.saveAs("#{dir}/in.osw")

      (backend || default_backend).execute(dir)

      model.setSqlFile(OpenStudio::SqlFile.new(OpenStudio::Path.new("#{dir}/run/eplusout.sql")))
      "#{dir}/run"
    end

    # @return [Boolean] completed with no Fatal and no Severe errors
    def clean_run?(run_dir)
      err = File.read("#{run_dir}/eplusout.err")
      err.include?('EnergyPlus Completed Successfully') &&
        !err.match?(/\*\*\s*Fatal\s*\*\*/) && !err.match?(/\*\* Severe  \*\*/)
    end

    # Annual (or run-period) site energy results from an attached SQL.
    # @return [Hash] kWh totals + end-use breakdown + floor area
    def energy_results(model)
      sql = model.sqlFile
      raise('no SQL attached — run run_energyplus! first') if sql.empty?

      sql = sql.get
      gj_to_kwh = 277.777778
      total_gj = optional(sql.totalSiteEnergy)
      results = {
        'total_site_kwh' => total_gj ? (total_gj * gj_to_kwh).round(1) : nil,
        'electricity_kwh' => scaled(sql.electricityTotalEndUses, gj_to_kwh),
        'natural_gas_kwh' => scaled(sql.naturalGasTotalEndUses, gj_to_kwh),
        'district_heating_kwh' => district(sql, %i[districtHeatingTotalEndUses districtHeatingWaterTotalEndUses], gj_to_kwh),
        'district_cooling_kwh' => district(sql, %i[districtCoolingTotalEndUses], gj_to_kwh),
        'end_uses_kwh' => {
          'heating' => end_use(sql, 'Heating', gj_to_kwh),
          'cooling' => end_use(sql, 'Cooling', gj_to_kwh),
          'fans' => end_use(sql, 'Fans', gj_to_kwh),
          'pumps' => end_use(sql, 'Pumps', gj_to_kwh),
          'interior_lighting' => end_use(sql, 'InteriorLighting', gj_to_kwh),
          'interior_equipment' => end_use(sql, 'InteriorEquipment', gj_to_kwh),
          'water_systems' => end_use(sql, 'WaterSystems', gj_to_kwh)
        }
      }
      area = model.getBuilding.floorArea
      results['floor_area_m2'] = area.round(1)
      results['eui_kwh_per_m2'] = (results['total_site_kwh'] / area).round(1) if results['total_site_kwh'] && area.positive?
      results
    end

    # Facility 'Time Setpoint Not Met During Occupied' hours (SystemSummary).
    # @return [Hash] { 'heating' => Float|nil, 'cooling' => Float|nil }
    def unmet_occupied_hours(model)
      sql = model.sqlFile
      return { 'heating' => nil, 'cooling' => nil } if sql.empty?

      sql = sql.get
      query = lambda do |column|
        value = sql.execAndReturnFirstDouble(
          "SELECT Value FROM TabularDataWithStrings WHERE ReportName='SystemSummary' " \
          "AND TableName='Time Setpoint Not Met' AND RowName='Facility' AND ColumnName='#{column}'"
        )
        value.is_initialized ? value.get : nil
      end
      { 'heating' => query.call('During Occupied Heating'),
        'cooling' => query.call('During Occupied Cooling') }
    end

    # Per-zone 'Time Setpoint Not Met During Occupied' hours (SystemSummary) —
    # the per-thermal-block resolution behind NECB 8.4.1.2.(3)/(4), which are
    # written "for each thermal block". Row names are EnergyPlus upper-cased
    # zone names; callers matching against model zones should compare
    # case-insensitively. Empty hash when no SQL is attached.
    # @return [Hash{String=>Hash}] { 'ZONE NAME' => { 'heating' => Float, 'cooling' => Float } }
    def zone_unmet_occupied_hours(model)
      sql = model.sqlFile
      return {} if sql.empty?

      sql = sql.get
      zones = {}
      { 'heating' => 'During Occupied Heating', 'cooling' => 'During Occupied Cooling' }.each do |metric, column|
        rows = sql.execAndReturnVectorOfString(
          "SELECT RowName || '|' || Value FROM TabularDataWithStrings WHERE ReportName='SystemSummary' " \
          "AND TableName='Time Setpoint Not Met' AND ColumnName='#{column}' AND RowName <> 'Facility'"
        )
        next unless rows.is_initialized

        rows.get.each do |line|
          name, _, value = line.rpartition('|')
          next if name.empty?

          (zones[name] ||= {})[metric] = value.to_f
        end
      end
      zones
    end

    # Request output variables at RunPeriod frequency (key '*'), idempotently —
    # the NECB 8.4.4.13.(2)(g) auxiliary-fuel election (D-52) needs per-equipment
    # heating energy from the proposed annual run.
    def request_run_period_variables!(model, names)
      existing = model.getOutputVariables.map { |v| [v.keyValue, v.variableName] }
      names.each do |name|
        next if existing.include?(['*', name])

        variable = OpenStudio::Model::OutputVariable.new(name, model)
        variable.setKeyValue('*')
        variable.setReportingFrequency('RunPeriod')
      end
    end

    # Sum a reported variable per KeyValue over the WEATHER run period(s) only.
    # EnvironmentType = 3 filters out design days — a shared DDY can carry
    # dozens, and an unfiltered sum silently mixes them in (the D-56 analysis
    # trap). Keys are EnergyPlus UPPER-CASED object names.
    # @return [Hash{String=>Float}] { 'COIL NAME' => joules }
    def run_period_sums(model, variable_name)
      sql = model.sqlFile
      return {} if sql.empty?

      rows = sql.get.execAndReturnVectorOfString(
        "SELECT d.KeyValue || '|' || SUM(r.Value) FROM ReportData r " \
        'JOIN ReportDataDictionary d ON r.ReportDataDictionaryIndex = d.ReportDataDictionaryIndex ' \
        'JOIN Time t ON r.TimeIndex = t.TimeIndex ' \
        "WHERE d.Name = '#{variable_name}' AND t.EnvironmentPeriodIndex IN " \
        '(SELECT EnvironmentPeriodIndex FROM EnvironmentPeriods WHERE EnvironmentType = 3) ' \
        'GROUP BY d.KeyValue'
      )
      return {} unless rows.is_initialized

      rows.get.to_h do |line|
        name, _, value = line.rpartition('|')
        [name, value.to_f]
      end
    end

    def optional(value)
      value.is_initialized ? value.get : nil
    end

    def scaled(value, factor)
      v = optional(value)
      v.nil? ? nil : (v * factor).round(1)
    end

    # District end-use SqlFile accessors were renamed across OpenStudio versions.
    def district(sql, method_names, factor)
      name = method_names.find { |m| sql.respond_to?(m) }
      name ? scaled(sql.public_send(name), factor) : nil
    end

    # SqlFile end-use accessors are per-fuel (electricityHeating, naturalGasHeating,
    # ...) — sum the category across every fuel from the tabular End Uses table.
    def end_use(sql, category, factor)
      value = sql.execAndReturnFirstDouble(
        "SELECT SUM(Value) FROM TabularDataWithStrings " \
        "WHERE ReportName='AnnualBuildingUtilityPerformanceSummary' AND TableName='End Uses' " \
        "AND RowName='#{end_use_row(category)}' AND Units='GJ'"
      )
      value.is_initialized ? (value.get * factor).round(1) : nil
    end

    def end_use_row(category)
      { 'Heating' => 'Heating', 'Cooling' => 'Cooling', 'Fans' => 'Fans', 'Pumps' => 'Pumps',
        'InteriorLighting' => 'Interior Lighting', 'InteriorEquipment' => 'Interior Equipment',
        'WaterSystems' => 'Water Systems' }.fetch(category)
    end
  end
end

File: python/btap/simulation/runner.py
"""Simulation execution, backend-agnostic (port of btap-simulation's
runner.rb): attach weather, prepare the run directory (sizing flags, run
period, in.osm + in.osw), delegate the EnergyPlus invocation to a Backend,
and parse results from the attached SQL.

Port renames (D-79 protocol): run_energyplus! -> run_energyplus,
clean_run? -> is_clean_run, attach_weather! -> attach_weather,
request_run_period_variables! -> request_run_period_variables. The
openstudio_cli probe is gone — the Python Local backend runs the provisioned
engine, not the CLI.
"""

from __future__ import annotations

from pathlib import Path

from btap._compat import opt, ruby_round

_default_backend = None


def default_backend():
    """The process-wide default backend. The umbrella calls run_energyplus at
    ~8 sites without a backend argument — this is how a --backend flag
    reaches them all without threading a parameter through every phase. Set
    once at startup; a per-call backend= still wins. Not thread-safe by
    design; forked children set their own."""
    global _default_backend
    if _default_backend is None:
        from btap.simulation.backends import Local
        _default_backend = Local()
    return _default_backend


def set_default_backend(backend):
    global _default_backend
    _default_backend = backend


def attach_weather(model, *, epw, ddy):
    """Attach an EPW + its design days (required before any sizing run).

    Design days are REPLACED, not appended (a model already carrying design
    days would otherwise size on duplicates), and filtered to the 99.6%
    heating / 0.4% cooling extremes — the legacy convention. A full DDY
    carries ~40 entries including monthly and shoulder-season cooling days;
    sizing plant equipment on those breaks E+'s cooling-tower UA autosizing.
    """
    import re

    import openstudio

    # Absolute path: the model's WeatherFile object stores this string and it
    # is later resolved from the RUN directory, not the caller's.
    epw_file = openstudio.EpwFile(openstudio.path(str(Path(epw).resolve())))
    openstudio.model.WeatherFile.setWeatherFile(model, epw_file)
    workspace = opt(openstudio.energyplus.loadAndTranslateIdf(str(ddy)))
    if workspace is None:
        raise ValueError(f"could not parse design days from {ddy}")

    for dd in model.getDesignDays():
        dd.remove()
    # the legacy default list — NOT a bare /.4%/, which would also pull the
    # MONTHLY .4% days (a January cooling day's ~2C wet-bulb breaks the
    # tower UA solve). Ruby's =~ is a SEARCH: re.search, never match.
    keep = [r"Htg 99.6. Condns DB", r"Clg .4% Condns DB=>MWB",
            r"Clg 0.4% Condns DB=>MCWB", r"Clg .4. Condns WB=>MDB"]
    all_days = list(workspace.getDesignDays())
    extremes = [dd for dd in all_days
                if any(re.search(p, dd.nameString()) for p in keep)]
    if not extremes:  # odd DDY: keep everything rather than none
        extremes = all_days
    for dd in extremes:
        model.addObject(dd.clone())
    if len(model.getDesignDays()) == 0:
        raise ValueError(f"no design days found in {ddy}")
    return model


def run_energyplus(model, run_dir, *, sizing_only=False, run_period=None, backend=None):
    """Prepare the run directory, run EnergyPlus via the chosen backend, and
    attach the result SQL to the model. Returns the run output directory
    (containing eplusout.err / eplusout.sql).

    run_period: {'begin_month':, 'begin_day':, 'end_month':, 'end_day':} —
    override for the weather run (tests use one week; compliance is annual).
    """
    import openstudio

    run_dir = Path(run_dir)
    run_dir.mkdir(parents=True, exist_ok=True)
    sim = model.getSimulationControl()
    sim.setDoZoneSizingCalculation(True)
    sim.setDoSystemSizingCalculation(True)
    sim.setDoPlantSizingCalculation(True)
    sim.setRunSimulationforSizingPeriods(True)
    sim.setRunSimulationforWeatherFileRunPeriods(not sizing_only)
    if not sizing_only and run_period:
        rp = model.getRunPeriod()
        rp.setBeginMonth(run_period["begin_month"])
        rp.setBeginDayOfMonth(run_period["begin_day"])
        rp.setEndMonth(run_period["end_month"])
        rp.setEndDayOfMonth(run_period["end_day"])
    model.save(openstudio.path(str(run_dir / "in.osm")), True)
    # in.osw kept for run-dir artifact parity with the Ruby gem (side-by-side
    # debugging of Leg-B corpus runs); the Python Local backend itself reads
    # only in.osm.
    osw = openstudio.WorkflowJSON()
    osw.setSeedFile(openstudio.path(str((run_dir / "in.osm").resolve())))
    osw.saveAs(openstudio.path(str(run_dir / "in.osw")))

    (backend or default_backend()).execute(str(run_dir))

    model.setSqlFile(openstudio.SqlFile(openstudio.path(str(run_dir / "run" / "eplusout.sql"))))
    return str(run_dir / "run")


def is_clean_run(run_out_dir) -> bool:
    """Completed with no Fatal and no Severe errors."""
    import re

    err = (Path(run_out_dir) / "eplusout.err").read_text(encoding="utf-8", errors="replace")
    return ("EnergyPlus Completed Successfully" in err
            and not re.search(r"\*\*\s*Fatal\s*\*\*", err)
            and not re.search(r"\*\* Severe  \*\*", err))


GJ_TO_KWH = 277.777778


def energy_results(model) -> dict:
    """Annual (or run-period) site energy results from an attached SQL:
    kWh totals + end-use breakdown + floor area. Same keys, same rounding
    (ruby_round — these values land in report.json, which Leg B diffs)."""
    sql = opt(model.sqlFile())
    if sql is None:
        raise RuntimeError("no SQL attached — run run_energyplus first")

    total_gj = opt(sql.totalSiteEnergy())
    results = {
        "total_site_kwh": ruby_round(total_gj * GJ_TO_KWH, 1) if total_gj is not None else None,
        "electricity_kwh": _scaled(sql.electricityTotalEndUses()),
        "natural_gas_kwh": _scaled(sql.naturalGasTotalEndUses()),
        "district_heating_kwh": _district(sql, ["districtHeatingTotalEndUses",
                                                "districtHeatingWaterTotalEndUses"]),
        "district_cooling_kwh": _district(sql, ["districtCoolingTotalEndUses"]),
        "end_uses_kwh": {
            "heating": _end_use(sql, "Heating"),
            "cooling": _end_use(sql, "Cooling"),
            "fans": _end_use(sql, "Fans"),
            "pumps": _end_use(sql, "Pumps"),
            "interior_lighting": _end_use(sql, "InteriorLighting"),
            "interior_equipment": _end_use(sql, "InteriorEquipment"),
            "water_systems": _end_use(sql, "WaterSystems"),
        },
    }
    area = model.getBuilding().floorArea()
    results["floor_area_m2"] = ruby_round(area, 1)
    if results["total_site_kwh"] is not None and area > 0:
        results["eui_kwh_per_m2"] = ruby_round(results["total_site_kwh"] / area, 1)
    return results


def unmet_occupied_hours(model) -> dict:
    """Facility 'Time Setpoint Not Met During Occupied' hours (SystemSummary).
    Values are RAW SQL floats, never rounded (the Leg-B spec carries a
    tolerance for exactly this field)."""
    sql = opt(model.sqlFile())
    if sql is None:
        return {"heating": None, "cooling": None}

    def query(column):
        return opt(sql.execAndReturnFirstDouble(
            "SELECT Value FROM TabularDataWithStrings WHERE ReportName='SystemSummary' "
            f"AND TableName='Time Setpoint Not Met' AND RowName='Facility' AND ColumnName='{column}'"
        ))

    return {"heating": query("During Occupied Heating"),
            "cooling": query("During Occupied Cooling")}


def zone_unmet_occupied_hours(model) -> dict:
    """Per-zone 'Time Setpoint Not Met During Occupied' hours — the
    per-thermal-block resolution behind NECB 8.4.1.2.(3)/(4). Row names are
    EnergyPlus UPPER-CASED zone names; match case-insensitively. Empty dict
    when no SQL is attached."""
    sql = opt(model.sqlFile())
    if sql is None:
        return {}

    zones: dict = {}
    for metric, column in (("heating", "During Occupied Heating"),
                           ("cooling", "During Occupied Cooling")):
        rows = opt(sql.execAndReturnVectorOfString(
            "SELECT RowName || '|' || Value FROM TabularDataWithStrings "
            "WHERE ReportName='SystemSummary' AND TableName='Time Setpoint Not Met' "
            f"AND ColumnName='{column}' AND RowName <> 'Facility'"
        ))
        if rows is None:
            continue
        for line in rows:
            name, _, value = line.rpartition("|")
            if not name:
                continue
            zones.setdefault(name, {})[metric] = _to_f(value)
    return zones


def request_run_period_variables(model, names):
    """Request output variables at RunPeriod frequency (key '*'),
    idempotently — the NECB 8.4.4.13.(2)(g) auxiliary-fuel election (D-52)
    needs per-equipment heating energy from the proposed annual run."""
    import openstudio

    existing = [(v.keyValue(), v.variableName()) for v in model.getOutputVariables()]
    for name in names:
        if ("*", name) in existing:
            continue
        variable = openstudio.model.OutputVariable(name, model)
        variable.setKeyValue("*")
        variable.setReportingFrequency("RunPeriod")


def run_period_sums(model, variable_name) -> dict:
    """Sum a reported variable per KeyValue over the WEATHER run period(s)
    only. EnvironmentType = 3 filters out design days — a shared DDY can
    carry dozens, and an unfiltered sum silently mixes them in (the D-56
    trap). Keys are EnergyPlus UPPER-CASED object names; values joules."""
    sql = opt(model.sqlFile())
    if sql is None:
        return {}

    rows = opt(sql.execAndReturnVectorOfString(
        "SELECT d.KeyValue || '|' || SUM(r.Value) FROM ReportData r "
        "JOIN ReportDataDictionary d ON r.ReportDataDictionaryIndex = d.ReportDataDictionaryIndex "
        "JOIN Time t ON r.TimeIndex = t.TimeIndex "
        f"WHERE d.Name = '{variable_name}' AND t.EnvironmentPeriodIndex IN "
        "(SELECT EnvironmentPeriodIndex FROM EnvironmentPeriods WHERE EnvironmentType = 3) "
        "GROUP BY d.KeyValue"
    ))
    if rows is None:
        return {}
    result = {}
    for line in rows:
        name, _, value = line.rpartition("|")
        result[name] = _to_f(value)
    return result


def _to_f(value: str) -> float:
    """Ruby String#to_f: best-effort parse, 0.0 on garbage — SQL join output
    is numeric-or-empty and Ruby's leniency is part of the ported contract."""
    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0


def _scaled(optional_gj):
    value = opt(optional_gj)
    return None if value is None else ruby_round(value * GJ_TO_KWH, 1)


def _district(sql, method_names):
    """District end-use SqlFile accessors were renamed across OpenStudio
    versions — probe with hasattr (the respond_to? port)."""
    for name in method_names:
        if hasattr(sql, name):
            return _scaled(getattr(sql, name)())
    return None


def _end_use(sql, category):
    """SqlFile end-use accessors are per-fuel — sum the category across every
    fuel from the tabular End Uses table."""
    row = {"Heating": "Heating", "Cooling": "Cooling", "Fans": "Fans", "Pumps": "Pumps",
           "InteriorLighting": "Interior Lighting", "InteriorEquipment": "Interior Equipment",
           "WaterSystems": "Water Systems"}[category]
    value = opt(sql.execAndReturnFirstDouble(
        "SELECT SUM(Value) FROM TabularDataWithStrings "
        "WHERE ReportName='AnnualBuildingUtilityPerformanceSummary' AND TableName='End Uses' "
        f"AND RowName='{row}' AND Units='GJ'"
    ))
    return None if value is None else ruby_round(value * GJ_TO_KWH, 1)
